Add Vulns to `os_versions` API

mostlikelee opened the pull request:
#15797

getvictor commented on ee/server/service/hosts.go:
```diff
@@ -19,3 +21,105 @@ func (svc *Service) HostByIdentifier(ctx context.Context, identifier string, opt
+		}
+	}
+
+	if opts.OrderKey == "hosts_count" && opts.OrderDirection == fleet.OrderAscending {
```
Is `hosts_count` the only allowed order key? If so, please add validation if needed.

getvictor commented on ee/server/service/hosts.go:
```diff
@@ -19,3 +21,105 @@ func (svc *Service) HostByIdentifier(ctx context.Context, identifier string, opt
+
+	count = len(osVersions.OSVersions)
+
+	osVersions.OSVersions = paginateOSVersions(osVersions.OSVersions, opts)
```
This seems inefficient to paginate here. Any way to reduce DB calls when pagination is needed?

mostlikelee replied:
`ds.OSVersions` returns a single row from the  `aggregated_stats` including a JSON field that is unmarshaled and returned.  I'm not convinced pagination here gives much benefit.

getvictor commented on ee/server/service/hosts.go:
```diff
@@ -19,3 +21,105 @@ func (svc *Service) HostByIdentifier(ctx context.Context, identifier string, opt
+	}
+
+	if name != nil && version == nil {
+		return nil, count, nil, &fleet.BadRequestError{Message: "Cannot specify os_name without os_version"}
```
Please add validation tests for these errors.

getvictor commented on ee/server/service/hosts.go:
```diff
@@ -19,3 +21,105 @@ func (svc *Service) HostByIdentifier(ctx context.Context, identifier string, opt
+		// differentiate case where team was added after UpdateOSVersions last ran
+		if teamID != nil && *teamID > 0 {
+			// most of the time, team should exist so checking here saves unnecessary db calls
+			_, err := svc.ds.Team(ctx, *teamID)
```
Please add a test for this.

getvictor commented on ee/server/service/hosts.go:
```diff
@@ -19,3 +21,105 @@ func (svc *Service) HostByIdentifier(ctx context.Context, identifier string, opt
+
+	var metaData *fleet.PaginationMetadata
+	if opts.IncludeMetadata {
+		metaData = &fleet.PaginationMetadata{HasPreviousResults: opts.Page > 0}
```
Please add a test for metadata

getvictor commented on server/service/integration_enterprise_test.go:
```diff
@@ -3085,6 +3085,79 @@ func (s *integrationEnterpriseTestSuite) TestListHosts() {
+func (s *integrationEnterpriseTestSuite) TestOSVersions() {
+	t := s.T()
+
+	testOS := fleet.OperatingSystem{Name: "Windows 11 Pro", Version: "10.0.22621.2861", Arch: "x86_64", KernelVersion: "10.0.22621.2861", Platform: "windows"}
```
Should we test for macOS as well?

getvictor commented on server/service/hosts.go:
```diff
@@ -1789,16 +1796,85 @@ func (svc *Service) OSVersions(ctx context.Context, teamID *uint, platform *stri
+	return osVersions, count, metaData, nil
+}
+
+func paginateOSVersions(slice []fleet.OSVersion, opts fleet.ListOptions) []fleet.OSVersion {
```
Isn't this the same as the ee version? If so, can we use the same function?

getvictor commented on server/service/hosts.go:
```diff
@@ -1789,16 +1796,85 @@ func (svc *Service) OSVersions(ctx context.Context, teamID *uint, platform *stri
+		}
+	}
+
+	if opts.OrderKey == "hosts_count" && opts.OrderDirection == fleet.OrderAscending {
```
I see a lot of similarities between this and the ee version. Can they share some of the code in this method?

getvictor commented on server/service/hosts.go:
```diff
@@ -1741,45 +1742,51 @@ func hostsReportEndpoint(ctx context.Context, request interface{}, svc fleet.Ser
 
 	if name != nil && version == nil {
-		return nil, &fleet.BadRequestError{Message: "Cannot specify os_name without os_version"}
+		return nil, count, nil, &fleet.BadRequestError{Message: "Cannot specify os_name without os_version"}
```
Tests needed for this validation. May be easier if the same validation logic is shared with ee version.

getvictor reviewed (changes requested):
A few comments about testing and potential refactoring.

mostlikelee commented:
@getvictor Thanks for the review!  I added most tests to `integration_core` as I refactored the `ee` implementation to reuse the same version, adding a boolean param.  `integration_enterprise` is only testing the additional CVE metadata as well as the team related return values.  Curious what you think.

I'm of the opinion that pagination isn't giving us much benefit here, but since the frontend already has that implemented, we could save that for a future refactor.

mostlikelee commented on server/service/hosts.go:
```diff
@@ -1789,16 +1796,87 @@ func (svc *Service) OSVersions(ctx context.Context, teamID *uint, platform *stri
+		}
+	}
+
+	if opts.OrderKey != "" && opts.OrderKey != "hosts_count" {
+		return nil, count, nil, &fleet.BadRequestError{Message: "Invalid order key"}
+	}
```
good call, i'll do that real quick

getvictor reviewed (approved):
LGTM
